Allow specifying vitamins by weight (mass) in items

ehughsbaird opened the pull request:
#### Summary
Infrastructure "Allow specifying vitamins by weight (mass) in items"

#### Purpose of change
Fixes https://github.com/CleverRaven/Cataclysm-DDA/issues/64722

Often, nutrition info will list vitamins by weight. Additionally, RDA, being defined by humans, is subject to change. For these reasons, it should be possible to specify vitamins by weight in items definitions.

#### Describe the solution
To do this, vitamins need to know how weight corresponds to units, and items need to load in weight as well as units, and convert to units as necessary. Unfortunately, as items may be loaded before vitamins, both the loaded weight values and units values must be stored until all data is loaded and they can be converted to units. To facilitate this, turn the vitamins data stored in nutrients into a map from vitamin id to a union of int and units::mass. This can then be finalized after loading to convert all units::mass values to units.

Several functions were added to ease the difficulty in navigating this new structure, and safety checks were added to prevent bugs.

#### Describe alternatives you've considered
There's a lot of rounding that the whole vitamins system is subject to. This adds some more rounding: Any fractional vitamins due to weight are discarded. However, this is the same as the previous approach, but happens at load time instead of when the content is being added.

Additionally, vitamins (like iron) that have an RDA of less than 1mg cannot really benefit from this system.

#### Testing
Before:
|drink|shake|fortified shake|
|-|-|-|
|![Image of protein drink item description, showing 6% RDA Calcium](https://github.com/CleverRaven/Cataclysm-DDA/assets/44244083/366c8580-d3f8-4c84-8ba7-0c634ff0392c)|![Image of protein shake item description, showing 6% RDA calcium and 20% RDA vitamin C](https://github.com/CleverRaven/Cataclysm-DDA/assets/44244083/e2a080dc-5ff7-4d8e-9ad4-af708b59a5d4)|![Image of fortified protein shake item description, showing 11% RDA calcium, 4% RDA iron, and 25% RDA vitamin C](https://github.com/CleverRaven/Cataclysm-DDA/assets/44244083/43a450b5-6315-4bd2-8f04-dfefd70f2eb4)|

After:
|drink|shake|fortified shake|
|-|-|-|
|![Image of protein drink item description, showing 6% RDA Calcium](https://github.com/CleverRaven/Cataclysm-DDA/assets/44244083/7beda0e0-5ce6-49e8-9c53-6b8376bbd207)|![Image of protein shake item description, showing 6% RDA calcium and 20% RDA vitamin C](https://github.com/CleverRaven/Cataclysm-DDA/assets/44244083/2ddb1ed1-9620-483e-a06b-19546d9d20fd)|![Image of fortified protein shake item description, showing 11% RDA calcium, 4% RDA iron, and 25% RDA vitamin C](https://github.com/CleverRaven/Cataclysm-DDA/assets/44244083/24cf2155-5d17-4bd6-a6a1-915401f231fe)|

#### Additional context
I'd prefer the commits not be squashed here, if possible.

If anyone has any ideas on how to handle iron with this, I'd appreciate hearing them. ~~My only idea so far was to increase the precision of the vitamin (e.g. consume 9600 units per day, instead of 96).~~

Karol1223 commented on data/json/vitamin.json:
```diff
@@ -6,6 +6,11 @@
     "name": { "str": "Calcium" },
     "min": -48000,
     "max": 0,
+    "//": [
+      "96 units are consumed per day, with RDA of 1000mg / 96 = 10.42mg",
+      "1000mg for 19-50yr olds from https://ods.od.nih.gov/factsheets/calcium-HealthProfessional/#h2"
+    ],
+    "weight_per_unit": "10 mg",
```
As per https://www.fda.gov/food/nutrition-facts-label/daily-value-nutrition-and-supplement-facts-labels, the RDA for Calcium is 1300mg. I am not really sure which of these sources is more trustworthy, though, so I really just wanted to throw this out here as something to think about. Technically the FDA has been updated a year later, but I'm not 100% sure regardless...

Other than that, thank you a lot for doing this!

ehughsbaird replied:
The source I used has a table, I'll reproduce it here:


Age | Male | Female | Pregnant | Lactating
-- | -- | -- | -- | --
0–6 months* | 200 mg | 200 mg |   |  
7–12 months* | 260 mg | 260 mg |   |  
1–3 years | 700 mg | 700 mg |   |  
4–8 years | 1,000 mg | 1,000 mg |   |  
9–13 years | 1,300 mg | 1,300 mg |   |  
14–18 years | 1,300 mg | 1,300 mg | 1,300 mg | 1,300 mg
19–50 years | 1,000 mg | 1,000 mg | 1,000 mg | 1,000 mg
51–70 years | 1,000 mg | 1,200 mg |   |  
|>70+ years | 1,200 mg | 1,200 mg |  

It looks like the FDA is just reproducing the highest  (non-pregnant/lactating) value on the table (this is how it is with the other values). I figured most DDA characters are in the 19-50 age range, so that value would be more appropriate.

Karol1223 replied:
Yep, that seems reasonable to me. Apologies for the bother, and thank you again for taking the time to do this!

ehughsbaird committed:
Privatize nutrients::vitamins

In preparation for changing how vitamins are represented/stored under
the hood, privatize the structure storing vitamins and update the code
to interface with it correctly.

The new members added to the struct are the functions
set_vitamin(), add_vitamin(), remove_vitamin() and vitamins()
The other change is to move the vitamins variable to a private variable
vitamins_

Astyle reformatted the struct with the addition of the visibility

ehughsbaird committed:
Allow specifying vitamins in items by weight/mass

Often, nutrition info will list vitamins by weight. Additionally, RDA,
being defined by humans, is subject to change. For these reasons, it
should be possible to specify vitamins by weight in item definitions.

To do this, vitamins need to know how weight corresponds to units, and
items need to load in weight as well as units, and convert to units as
necessary. Unfortunately, as items may be loaded before vitamins, both
the loaded weight values and units values must be stored until all data
is loaded and they can be converted to units. To facilitate this, turn
the vitamins data stored in nutrients into a map from vitamin id to a
union of int and units::mass. This can then be finalized after loading
to convert all units::mass values to units.

Some nutrients functions modified to safely handle the union type.

LIMITATIONS:
There's a lot of rounding that the whole vitamins system is subject to.
This adds some more rounding: Any fractional vitamins due to weight are
discarded. However, this is the same as the previous approach, but
happens at load time instead of when the content is being added.
Additionally, vitamins (like iron) that have an RDA of less than 1mg
cannot really benefit from this system.

Demonstrate this by adding appropriate weights to nutrition vitamins
(barring iron) and modifying some items to specify their vitamins by
weight (both normally and through relative).
